### ISSUE TITLE
Allow extensions to specify a preferred scope (eg test)

### ISSUE THREAD

**Issue by holly-cummins** [quarkusio/quarkus#34021]
### Description

I've noticed that we don't have a clean way to add test-scoped extensions using `quarkus ext add`, so people have to edit the pom directly, which @maxandersen tells me is definitely not the preferred way. :)

We could do either of the following

- Update quarkus ext add to add either a --test flag or a --scope=test option. The first one is shorter, but potentially confusing because it could mean something like a dry run flag.
- Add a new metadata in quarkus-extension.yaml which defines a preferred scope (maybe preferred-scope: test) that defaults to compile. Then the `ext add` command can use this metadata to set the right scope.
 
Of those two, the metadata has less opportunity for user error, and it lets us display the correct command on http://extensions.quarkus.io.

I noticed this for the pact extension, but some other extensions would also benefit from this, such as `quarkus-jacoco` and `quarkus-junit5-mockito` and `quarkus-junit5-component`. 

@aloubyansky says 

"We could consider configuring quarkus launch modes instead: prod, test, dev and then map that to the corresponding maven and Gradle classpath. Otherwise we may need to configure the target classpath/scope for each build system in the extension metadata. I can imagine there will be more requests for dev mode only dependencies if we go with launch modes, which would be easy in case of Gradle."

The dev mode case is a bit complex, so I will start with test scope, since there's a clean mapping between launch mode and maven/gradle scope.

**Comment by holly-cummins** [comment]
The "users pass a scope on `ext add`" case is covered by #17539.

**Comment by holly-cummins** [comment]
We should also make sure not to lose the idea of dev mode dependencies, since they're a not-uncommon request. See, for example, the discussion on #30317. H2 would often be wanted in dev mode but not in production.

**Comment by holly-cummins** [comment]
The idea of dev mode dependencies is also an inverse of https://github.com/quarkusio/quarkus/issues/23175.

**Comment by holly-cummins** [comment]
Editing the title and description to adjust the scope to the scenario Alexey describes, and using #17539 to cover build-file scope of extension.

**Comment by aloubyansky** [comment]
> The idea of dev mode dependencies is also an inverse of #23175.

Not exactly, the way this looks to me. Disabling features does not imply changing the classpath. Launch modes, otoh, pretty much imply different project dependencies and as a consequence - the classpath.
